src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexSearcherHolder.cs

```diff
@@ -9,14 +9,21 @@ namespace Raven.Server.Documents.Indexes.Persistence.Lucene
 {
     public class IndexSearcherHolder
     {
+        private readonly Func<IndexSearcher> _recreateSearcher;
+
         private static readonly ILog Log = LogManager.GetLogger(typeof(IndexSearcherHolder));
 
         private volatile IndexSearcherHoldingState _current;
-        
-        public ManualResetEvent SetIndexSearcher(IndexSearcher searcher, bool wait)
+
+        public IndexSearcherHolder(Func<IndexSearcher> recreateSearcher)
+        {
```

Comment:
@ayende Would appreciate a review of this approach. It seems that we (in 3.x) might be creating a lot of readers in vain, because we are creating them after each Index commit and recreation is taking quite a long time even when they are not used by queries. I was thinking that we can create them on _first_ query request. Thoughts?